qiskit/transpiler/passes/layout/sabre_layout.py

```diff
@@ -227,59 +230,84 @@ def run(self, dag):
+
+        # At this point, we become a transformation pass, and apply the layout and the routing to
+        # the DAG directly.  This includes filling in the `property_set` data of the embed passes.
+
+        dag = self._ancilla_allocation_no_pass_manager(dag)
+        # The ancilla-allocation pass has expanded this since we set it above.
+        full_initial_layout = self.property_set["layout"]
+
+        # Set up a physical DAG to apply the Sabre result onto.  We do not need to run the
+        # `ApplyLayout` transpiler pass (which usually does this step), because we're about to apply
+        # the layout and routing together as part of resolving the Sabre result.
+        physical_qubits = QuantumRegister(self.coupling_map.size(), "q")
+        mapped_dag = DAGCircuit()
+        mapped_dag.add_qreg(physical_qubits)
+        mapped_dag.add_clbits(dag.clbits)
+        for creg in dag.cregs.values():
+            mapped_dag.add_creg(creg)
+        mapped_dag._global_phase = dag._global_phase
+        self.property_set["original_qubit_indices"] = {
+            bit: index for index, bit in enumerate(dag.qubits)
+        }
         self.property_set["final_layout"] = Layout(
-            {dag.qubits[k]: v for (k, v) in final_layout_dict.items()}
+            {
+                mapped_dag.qubits[
+                    component.coupling_map.graph[initial]
+                ]: component.coupling_map.graph[final]
+                for component in components
+                for initial, final in enumerate(component.final_permutation)
+            }
         )
-        canonical_register = dag.qregs["q"]
-        original_layout = NLayout.generate_trivial_layout(self.coupling_map.size())
-        for (
-            _layout_dict,
-            _final_layout_dict,
-            component_map,
-            sabre_result,
-            circuit_to_dag_dict,
-            local_dag,
-        ) in layout_components:
-            _apply_sabre_result(
+        for component in components:
+            # Sabre routing still returns all its swaps as on virtual qubits, so we need to expand
+            # each component DAG with the virtual ancillas that were allocated to it, so the layout
+            # application can succeed.  This is the last thing we do with the component DAG, so it's
+            # ok for us to modify it.
+            component_size = component.coupling_map.size()
+            dag_size = component.dag.num_qubits()
+            if component_size > dag_size:
+                used_physical = {full_initial_layout[logic] for logic in component.dag.qubits}
+                component.dag.add_qubits(
+                    [
+                        full_initial_layout[component.coupling_map.graph[phys]]
+                        for phys in range(component.dag.num_qubits(), component_size)
+                        if component.coupling_map.graph[phys] not in used_physical
+                    ]
+                )
```

Comment:
This is fine for now, but something we might want to consider is that we can only get to this point in the code if there is a single connected component. It's something that was an artifact from the development history of #9802, the original intent was we'd not need to route separately and we'd combine all the component results for the final out dag. But, it wasn't sound to do the swap insertion here and we had to apply the layout first and then re-route. So, I'm wondering if we can at least drop the for loop here if nothing else.

Reply:
We could drop the for loop, but all the component-specific handling is still necessary to handle the expansion of a virtual circuit to fill its physical component, so there's not a huge amount to gain from it.

The loop was in place already, so I'd just left it in this PR.

Reply:
Sure that makes sense, it was more just idle musing from me to remind myself to push up a PR after to remove the loop.